### ISSUE TITLE
[pacman] getting stuck on clang-aarch64

### ISSUE THREAD

**Issue by Wormnest** [msys2/MSYS2-packages#4340]
### Description / Steps to reproduce the issue

Since about a month GIMP's aarch64 CI runner is getting stuck when running `pacman --noconfirm -Suy`.
The last job that succeeded was Dec 11 and another from the same day and any later one is failing.

Most of the time (e.g. [here](https://gitlab.gnome.org/GNOME/gimp/-/jobs/3445501)) it already seems to stop before the databases are updated:
```
$ C:\msys64\usr\bin\bash -lc "bash -x ./build/windows/gitlab-ci/1_build-deps-msys2.sh"
+ set -e
+ [[ aarch64 == \a\a\r\c\h\6\4 ]]
+ export ARTIFACTS_SUFFIX=-a64
+ ARTIFACTS_SUFFIX=-a64
+ [[ CI_NATIVE != \C\I\_\N\A\T\I\V\E ]]
+ pacman --noconfirm -Suy
WARNING: Failed to terminate process: 1 error occurred:
	* failed to attach the runner process to the console of its parent process: The handle is invalid.
WARNING: Timed out waiting for the build to finish
```

[Sometimes](https://gitlab.gnome.org/GNOME/gimp/-/jobs/3441058) it gets a little further:
```
$ C:\msys64\usr\bin\bash -lc "bash -x ./build/windows/gitlab-ci/1_build-deps-msys2.sh"
+ set -e
+ [[ aarch64 == \a\a\r\c\h\6\4 ]]
+ export ARTIFACTS_SUFFIX=-a64
+ ARTIFACTS_SUFFIX=-a64
+ [[ CI_NATIVE != \C\I\_\N\A\T\I\V\E ]]
+ pacman --noconfirm -Suy
:: Synchronizing package databases...
 clangarm64 downloading...
 msys downloading...
:: Starting core system upgrade...
 there is nothing to do
:: Starting full system upgrade...
 there is nothing to do
+ pacman --noconfirm -S --needed base-devel mingw-w64-clang-aarch64-toolchain mingw-w64-clang-aarch64-meson mingw-w64-clang-aarch64-cairo mingw-w64-clang-aarch64-crt-git mingw-w64-clang-aarch64-glib-networking mingw-w64-clang-aarch64-gobject-introspection mingw-w64-clang-aarch64-json-glib mingw-w64-clang-aarch64-lcms2 mingw-w64-clang-aarch64-lensfun mingw-w64-clang-aarch64-libspiro mingw-w64-clang-aarch64-maxflow mingw-w64-clang-aarch64-openexr mingw-w64-clang-aarch64-pango mingw-w64-clang-aarch64-suitesparse mingw-w64-clang-aarch64-vala
WARNING: Failed to terminate process: 1 error occurred:
	* failed to attach the runner process to the console of its parent process: The handle is invalid.
```

When testing on my Ms Dev kit now it did not get stuck (but I do remember seeing that sometimes in the past). However, when checking with Process Explorer, I do see that after pacman closed the terminal, the pacman and conhost processes are still running.

### Expected behavior

Pacman finishes after doing its thing.

### Actual behavior

Pacman gets stuck

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

MSYS64_NT-10.0-22621

**Comment by Biswa96** [comment]
Does the CI script terminate msys2 processes after update? See https://github.com/msys2/setup-msys2/blob/8b0d40b8912601756301a7b3de7752d5dba969cd/main.js#L408

In that main.js file, `pacman -Syuu` updates base packages > terminate all msys2 related processes > update remaining packages.

**Comment by jeremyd2019** [comment]
There is also a bug (presumably in msys2-runtime), which I have never been able to debug, that manifests as processes hanging around when they should have exited.  This usually seems to happen when pacman uses gpgme to attempt to validate signatures.  As a workaround, I disable the validation of database signatures in pacman.conf, because the database signature verification seems to happen every time pacman is run, whereas package verification is only done when a package is being installed.  I still see occasional hangups in package verification though.

See also https://github.com/msys2/msys2-autobuild/issues/62 which I think is the closest thing to an existing bug tracking this.

**Comment by Wormnest** [comment]
> Does the CI script terminate msys2 processes after update?

We are now doing something [similar](https://gitlab.gnome.org/GNOME/gimp/-/merge_requests/1265), which seems to get us further, but it's finding a [database lock](https://gitlab.gnome.org/GNOME/gimp/-/jobs/3452833).

On my own Arm Dev Kit, I just updated pacman, which went without problems. However, it is now stuck in compiling part of GIMP (I think I've seen this before too). So, this getting stuck is probably not specific to pacman. Looking in process explorer, the innermost process is env.exe. Could it be related to reading/setting env vars, which I seem to remember can have problems from multiple threads.

**Comment by Biswa96** [comment]
The database lock file `/var/lib/pacman/db.lck` can be deleted before running pacman command. Though, I am not sure if that will fix the actual issue.

**Comment by Biswa96** [comment]
The msys2 related processes should be terminated outside of msys2 environment. For example, `taskkill` command in a batch file. See setup-msys2 repository as mentioned.

**Comment by brunolopesdsilv** [comment]
> The msys2 related processes should be terminated outside of msys2 environment. For example, `taskkill` command in a batch file. See setup-msys2 repository as mentioned.

I tried before with takkill but the exit code makes the job fail.
Inkscape uses taskkill, but they use the "retry" key in CI .yml.

**Comment by Biswa96** [comment]
OK, I am out of ideas then. By the way, @hmartinez82 has done some great work to port apps to aarch64. He may suggest some ideas.

**Comment by jeremyd2019** [comment]
I wish somebody with good knowledge of low-level debugging on WoA (and/or of Cygwin) could debug this, I have tried and had no luck (I always got an error getting the context of the main thread, from every debugger I tried: windbg, gdb, lldb).

**Comment by hmartinez82** [comment]
I even see this happening, randomly, in my personal laptop  when using pacman.

**Comment by Jehan** [comment]
Hi all! Now we have new runners contributed by Arm Ltd., additionally to the one by @hmartinez82. And this pacman getting stuck issue is also happening randomly on their runners.

Is there anything we could tell the admins at Arm to look for in order to help you debug this issue?

**Comment by hmartinez82** [comment]
@Jehan I'm glad they are having it too, so we now know it's not just my runner. I don't know what the issue is.

**Comment by jeremyd2019** [comment]
https://cygwin.com/pipermail/cygwin-patches/2024q1/012617.html

**Comment by Alovchin91** [comment]
Just a heads-up: this has been fixed by @jeremyd2019's [patch](https://github.com/msys2/msys2-runtime/pull/234) that made it into `msys2-runtime 3.5.4-5`.

**Comment by brunvonlope** [comment]
After the fix pacman is getting stuck on x64 and x86 MSYSTEMs at gpg phase using @lazka runners

**Comment by Alovchin91** [comment]
> After the fix pacman is getting stuck on x64 and x86 MSYSTEMs at gpg phase using @lazka runners

@jeremyd2019 Have you experienced this? 🤔

**Comment by Alovchin91** [comment]
There's this in `CancelSynchronousIo` function's [documentation](https://learn.microsoft.com/en-us/windows/win32/fileio/cancelsynchronousio-func):

> Most types of operations can be canceled immediately; other operations can continue toward completion before they are actually canceled and the caller is notified. The **CancelSynchronousIo** function does not wait for all canceled operations to complete.

I have seen some differences with other overlapped I/O functions between x64 and arm64 systems, namely `ReadDirectoryChangesEx` function has returned an error code synchronously on arm64, while on x64 it has (sometimes?) reported it to the overlapped operation.

Could it be that `CancelSynchronousIo` doesn't really cancel all the I/O on x64? I can't think of anything that could be wrong with your suspend thread solution.

**Comment by jeremyd2019** [comment]
> > After the fix pacman is getting stuck on x64 and x86 MSYSTEMs at gpg phase using @lazka runners
> 
> @jeremyd2019 Have you experienced this? 🤔

I have not, and I'm kind of confused by the report.  In this code, the MSYSTEM shouldn't have anything to do with it.  At least here, in pull requests and msys2-autobuild, pacman gets a lot of exercise and I haven't heard of any hangs...

Can you give details on the runners in questions (OS version at least)?  Are you able to investigate the runners manually, look at running processes and attach gdb/get backtraces of threads of hung process(es)?

**Comment by brunvonlope** [comment]
@jeremyd2019 The MSYSTEMs affected are clang64 and mingw32 (probably ucrt64, mingw64 and clang32 too). These two MSYSTEMs we use in x86 GIMP runners. clangarm64 is managed by ARM and it is working just fine after the fix.

The clang64 and mingw32 runners use Windows 10. Don't know NT build, @lazka should know

Here is an example of stuckness: https://gitlab.gnome.org/GNOME/gimp/-/issues/10782#note_2278116. I can't investigate them remotely.

**Comment by lazka** [comment]
We got a report on discord of things getting stuck now:

![image](https://github.com/user-attachments/assets/257a32d4-cd9f-478a-85eb-4cce4553c9b2)

**Comment by jeremyd2019** [comment]
Are there any steps to reproduce?  I made some effort to 'symbolicate' the stacktrace, it seems to be
```
0x1800461be = cygwait
0x1800f0547 = wait4
```
so it is waiting for a child process to exit, most likely.  Is the child process still running or not?

**Comment by Alovchin91** [comment]
There's a relatively new `cygwait` call here: https://github.com/msys2/MSYS2-packages/commit/f1898f02ce7a302f180f749bff55dc882709f4d8

Are you sure these hangs haven't been there in `msys2-runtime 3.5.4-2`?

**Comment by brunvonlope** [comment]
> Are you sure these hangs haven't been there in msys2-runtime 3.5.4-2?

@Alovchin91 No, unfortunately. It just coincided, as soon @Wormnest told me that the bug was fixed, x86 runners started to get  stuck.

**Comment by Alovchin91** [comment]
Hmm, I should have an old Intel MacBook somewhere, wonder if I'll be able to reproduce the hang 🤔 Pacman has been hanging consistently for me on arm64 before the fix.

**Comment by jeremyd2019** [comment]
We were also seeing some oddness from gpg itself around the time of its last update (a couple weeks ago).

**Comment by lazka** [comment]
Got one hang in CI just now: https://github.com/msys2/MINGW-packages/actions/runs/11962535058/job/33351156658?pr=22619

(should we open a new issue for this?)

**Comment by lazka** [comment]
Which reminds me that we had a stuck job in autobuild some days ago (2024-11-17): https://github.com/msys2/msys2-autobuild/actions/runs/11876484140/job/33094866915

stuck in a different place though

**Comment by jeremyd2019** [comment]
Probably.  I am going to suspect the CancelSynchronousIo call, without any further evidence.  It looks like it's hanging in pacman/gpgme so maybe I'll start a loop of pacman -Suu and see if it hangs so I can debug.

**Comment by hmartinez82** [comment]
If you think it's worth keeping the CancelSynchronousIo call only for ARM64, with the decision at runtime, IsWow64Process2 is an option.

**Comment by jeremyd2019** [comment]
I don't think it's necessary at all, the SuspendThread/GetThreadContext dance fixed the ARM64 issue, the CancelSynchronousIo was an attempt to avoid the necessity of TerminateThread at all.  If I can get some insight into what's going wrong, and it is due to CancelSyncronousIo somehow, it'd be better to just revert that part of the patch.

I sort of suspect whoever got that stack trace in the image above attached to the "wrong" process in the tree, though.  That stack looks like a process in the `waitpid` call in the "grandparent".  It could be that the wait_thread is messed up and not notifying it that the "parent" is gone, or it could more likely be that the "parent" is screwed somehow (that's what was happening with the longstanding ARM64 issue)

**Comment by jeremyd2019** [comment]
I was able to reproduce and poke in the debugger.  wait_thread is happily waiting in [`ReadFile`](https://github.com/msys2/msys2-runtime/blob/eafd9a22bc82cc443352a08c2853cd0d8649a436/winsup/cygwin/pinfo.cc#L1268), while the main thread is hanging in [`ForceCloseHandle1 (close_h, rd_proc_pipe);`](https://github.com/msys2/msys2-runtime/blob/eafd9a22bc82cc443352a08c2853cd0d8649a436/winsup/cygwin/pinfo.cc#L1397).  The [`pinfo::release` call](https://github.com/msys2/msys2-runtime/blob/eafd9a22bc82cc443352a08c2853cd0d8649a436/winsup/cygwin/sigproc.cc#L421) is *after* the wait thread should have been terminated, so it's apparent the `CancelSynchronousIo` didn't work for whatever reason (possibly due to canceling some other sychronous IO, letting it come back around in the loop and call `ReadFile` again.

**Comment by Alovchin91** [comment]
I'm astonished I guessed it right for once 😆

**Comment by Alovchin91** [comment]
Probably good to let @dscho  know as well

**Comment by Alovchin91** [comment]
> https://inbox.sourceware.org/cygwin-patches/765a746a-b5a1-cdab-242d-1ff3c5bd65ba@jdrake.com/T/#m98d5eebc0cda7df653275e3294abdcfd7cb9bf85

Probably makes sense to undo the `CancelSynchronousIo` change at least in `msys2-runtime`, since it has already been merged and published and is now causing issues, wdyt?

**Comment by jeremyd2019** [comment]
I did a quick commit last night (my time) and fired off some tests on x86_64 and ARM64 while I slept.
`while pacsift --base xorriso > /dev/null; do true; done` (this is how I reproduced the hang to debug).  

There were no hangs on either architecture, thought I saw a couple errors like this on ARM64:
`    111 [main] pacsift 29659 dofork: child -1 - forked process 6796 died unexpectedly, retry 0, exit code 0xC0000005, errno 11`

these didn't cause any other hang, error output, or abort the while loop...

**Comment by Alovchin91** [comment]
> `    111 [main] pacsift 29659 dofork: child -1 - forked process 6796 died unexpectedly, retry 0, exit code 0xC0000005, errno 11`

Some quick irresponsible googling has brought me [here](https://github.com/projectkudu/kudu/issues/2714) and [here](https://github.com/mintty/wsltty/issues/6).

tl;dr: Might have something to do with ASLR.

**Comment by jeremyd2019** [comment]
The fix for the x86_64 hang is merged now, and it looks like msys2-runtime 3.5.4-7 is already in the repo.

**Comment by jeremyd2019** [comment]
Looks like a hang on https://github.com/msys2/msys2-autobuild/actions/runs/11990401626/job/33427879361 not sure why though.

**Comment by Alovchin91** [comment]
Yes it looks like it goes like 1 package per 20 minutes or something.
